src/kernels/MIOpenMSELoss.cpp

```diff
@@ -0,0 +1,280 @@
+                                        const IO_TYPE* __restrict__ dO,
+                                        IO_TYPE* __restrict__ dI,
+                                        IO_TYPE* __restrict__ dT,
+                                        float divisor,
+                                        tensor_view_t<5> I_tv,
+                                        tensor_view_t<5> T_tv,
+                                        tensor_view_t<5> dO_tv,
+                                        tensor_view_t<5> dI_tv,
+                                        tensor_view_t<5> dT_tv)
+{
+    const size_t gid = blockIdx.x * blockDim.x + threadIdx.x;
+    size_t n0123 = gid / I_tv.size[4], n4 = gid % I_tv.size[4];
+    size_t n012 = n0123 / I_tv.size[3], n3 = n0123 % I_tv.size[3];
+    size_t n01 = n012 / I_tv.size[2], n2 = n012 % I_tv.size[2];
+    size_t n0 = n01 / I_tv.size[1], n1 = n01 % I_tv.size[1];
+
+    if(!(n0 < I_tv.size[0]))
+        return;
+
+    size_t Iidx = I_tv.get_tensor_view_idx({n0, n1, n2, n3, n4});
+    size_t Tidx = T_tv.get_tensor_view_idx({n0, n1, n2, n3, n4});
+
+    FLOAT_ACCUM iidxval  = CVT_FLOAT2ACCUM(I[Iidx]);
+    FLOAT_ACCUM tidxval  = CVT_FLOAT2ACCUM(T[Tidx]);
+    FLOAT_ACCUM dOidxval = CVT_FLOAT2ACCUM(dO[dO_tv.get_tensor_view_idx({dO_tv, 0})]);
+    FLOAT_ACCUM grad     = 2.0f * (iidxval - tidxval) / divisor * dOidxval;
+
+    if(dI != nullptr)
+    {
+        size_t dIidx = dI_tv.get_tensor_view_idx({n0, n1, n2, n3, n4});
+        dI[dIidx]    = CVT_ACCUM2FLOAT(grad);
+    }
+    if(dT != nullptr)
+    {
+        size_t dTidx = dT_tv.get_tensor_view_idx({n0, n1, n2, n3, n4});
+        dT[dTidx]    = CVT_ACCUM2FLOAT(-grad);
+    }
+}
+
+template <typename IO_TYPE>
+__device__ void DeviceMSELossUnreducedForward5d(const IO_TYPE* __restrict__ I,
+                                                const IO_TYPE* __restrict__ T,
+                                                IO_TYPE* __restrict__ O,
+                                                tensor_view_t<5> I_tv,
+                                                tensor_view_t<5> T_tv,
+                                                tensor_view_t<5> O_tv)
+{
+    const size_t gid = blockIdx.x * blockDim.x + threadIdx.x;
+
+    size_t n0123 = gid / I_tv.size[4], n4 = gid % I_tv.size[4];
+    size_t n012 = n0123 / I_tv.size[3], n3 = n0123 % I_tv.size[3];
+    size_t n01 = n012 / I_tv.size[2], n2 = n012 % I_tv.size[2];
+    size_t n0 = n01 / I_tv.size[1], n1 = n01 % I_tv.size[1];
+
+    if(!(n0 < I_tv.size[0]))
+        return;
+
+    size_t Iidx = I_tv.get_tensor_view_idx({n0, n1, n2, n3, n4});
+    size_t Tidx = T_tv.get_tensor_view_idx({n0, n1, n2, n3, n4});
+    size_t Oidx = O_tv.get_tensor_view_idx({n0, n1, n2, n3, n4});
+
+    FLOAT_ACCUM iidxval = CVT_FLOAT2ACCUM(I[Iidx]);
+    FLOAT_ACCUM tidxval = CVT_FLOAT2ACCUM(T[Tidx]);
+
+    FLOAT_ACCUM oidxval = (iidxval - tidxval) * (iidxval - tidxval);
+
+    O[Oidx] = CVT_ACCUM2FLOAT(oidxval);
+}
```

Comment:
It is absolutely identical to `DeviceMSELossForward5d`
Could you check this thread https://github.com/ROCm/MIOpen/pull/3166#discussion_r1705616950 and merge "reduce" and "unreduce" code in the kernel, driver and test.

Reply:
Oops, I should have dropped the Unreduced kernels entirely (given that it basically have no performance gains, and I’ve basically blocked all codepaths leading to it with `IsImprovementOverROCm` always return `false`).